From 6b95e790ca7c74faaa0ba8451a748e1f9a228326 Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Mon, 6 Sep 2021 00:12:04 +0900
Subject: [PATCH] nghttpx: Fix uninitialized dnf fields

---
 src/shrpx_config.h | 1 +
 src/shrpx_worker.h | 1 +
 2 files changed, 2 insertions(+)

diff --git a/src/shrpx_config.h b/src/shrpx_config.h
index e5a0fc6a..b27e891d 100644
--- a/src/shrpx_config.h
+++ b/src/shrpx_config.h
@@ -512,6 +512,7 @@ struct DownstreamAddrGroupConfig {
       : pattern(pattern),
         affinity{SessionAffinity::NONE},
         redirect_if_not_tls(false),
+        dnf{false},
         timeout{} {}
 
   StringRef pattern;
diff --git a/src/shrpx_worker.h b/src/shrpx_worker.h
index e736aad8..cd75bd63 100644
--- a/src/shrpx_worker.h
+++ b/src/shrpx_worker.h
@@ -191,6 +191,7 @@ struct SharedDownstreamAddr {
       : balloc(1024, 1024),
         affinity{SessionAffinity::NONE},
         redirect_if_not_tls{false},
+        dnf{false},
         timeout{} {}
 
   SharedDownstreamAddr(const SharedDownstreamAddr &) = delete;
-- 
2.25.1

