From 50a1121d812fc2091302cfb97364bafaee771f79 Mon Sep 17 00:00:00 2001
From: Tatsuhiro Tsujikawa <tatsuhiro.t@gmail.com>
Date: Mon, 8 Feb 2021 22:22:05 +0900
Subject: [PATCH] nghttpx: Remove trailing white space after $method log
 variable

---
 src/shrpx_log.cc | 1 -
 1 file changed, 1 deletion(-)

diff --git a/src/shrpx_log.cc b/src/shrpx_log.cc
index 3789c779..106aae7c 100644
--- a/src/shrpx_log.cc
+++ b/src/shrpx_log.cc
@@ -639,7 +639,6 @@ void upstream_accesslog(const std::vector<LogFragment> &lfv,
       break;
     case LogFragmentType::METHOD:
       std::tie(p, last) = copy(method, p, last);
-      std::tie(p, last) = copy(' ', p, last);
       break;
     case LogFragmentType::PATH:
       std::tie(p, last) = copy_escape(path, p, last);
-- 
2.25.1

