diff --git a/src/Makefile.am b/src/Makefile.am
index 03821648..d1fd78f2 100644
--- a/src/Makefile.am
+++ b/src/Makefile.am
@@ -49,17 +49,21 @@ AM_CPPFLAGS = \
 	@ZLIB_CFLAGS@ \
 	@DEFS@
 
+AM_LDFLAGS = -static-libtool-libs
+
 LDADD = $(top_builddir)/lib/libnghttp2.la \
 	$(top_builddir)/third-party/liburl-parser.la \
 	$(top_builddir)/third-party/libllhttp.la \
-	@JEMALLOC_LIBS@ \
+	${top_builddir}/third-party/libneverbleed.la \
+	/usr/lib/x86_64-linux-gnu/libjemalloc.a \
 	@LIBXML2_LIBS@ \
-	@LIBEV_LIBS@ \
-	@OPENSSL_LIBS@ \
-	@LIBCARES_LIBS@ \
+	/usr/lib/x86_64-linux-gnu/libev.a \
+	/usr/lib/x86_64-linux-gnu/libssl.a \
+	/usr/lib/x86_64-linux-gnu/libcrypto.a \
+	/usr/lib/x86_64-linux-gnu/libcares.a \
 	@SYSTEMD_LIBS@ \
 	@JANSSON_LIBS@ \
-	@ZLIB_LIBS@ \
+	/usr/lib/x86_64-linux-gnu/libz.a \
 	@APPLDFLAGS@
 
 if ENABLE_APP
@@ -169,7 +173,6 @@ endif # HAVE_MRUBY
 
 if HAVE_NEVERBLEED
 libnghttpx_a_CPPFLAGS += -I${top_srcdir}/third-party/neverbleed
-nghttpx_LDADD += ${top_builddir}/third-party/libneverbleed.la
 endif # HAVE_NEVERBLEED
 
 if HAVE_CUNIT
diff --git a/third-party/Makefile.am b/third-party/Makefile.am
index 59252362..73053d98 100644
--- a/third-party/Makefile.am
+++ b/third-party/Makefile.am
@@ -43,7 +43,6 @@ libllhttp_la_CPPFLAGS = -I${srcdir}/llhttp/include
 if HAVE_NEVERBLEED
 noinst_LTLIBRARIES += libneverbleed.la
 libneverbleed_la_CPPFLAGS = @OPENSSL_CFLAGS@
-libneverbleed_la_LIBADD = @OPENSSL_LIBS@
 libneverbleed_la_SOURCES = neverbleed/neverbleed.c neverbleed/neverbleed.h
 endif # HAVE_NEVERBLEED
 
